src/mlpack/methods/ann/ncf_impl.hpp

```diff
@@ -0,0 +1,254 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_NCF_IMPL_HPP
+#define MLPACK_METHODS_ANN_NCF_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "ncf.hpp"
+
+#include "visitor/forward_visitor.hpp"
+#include "visitor/backward_visitor.hpp"
+#include "visitor/deterministic_set_visitor.hpp"
+#include "visitor/gradient_set_visitor.hpp"
+#include "visitor/gradient_visitor.hpp"
+#include "visitor/set_input_height_visitor.hpp"
+#include "visitor/set_input_width_visitor.hpp"
+
+#include <boost/serialization/variant.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename Model, typename InitializationRuleType>
+NCFNetwork<Model, InitializationRuleType>::NCFNetwork(
+    Model& userModel, Model& itemModel,
+    InitializationRuleType initializeRule) :
+    userModel(userModel),
+    itemModel(itemModel),
+    initializeRule(std::move(initializeRule))
+{
+  /* Nothing to do here */
+}
+
+template<typename Model, typename InitializationRuleType>
+void NCFNetwork<Model, InitializationRuleType>::ResetData(arma::mat user,
+                                                          arma::mat item,
+                                                          arma::mat responses)
+{
+  numFunctions = responses.n_cols;
+  this->user = std::move(user);
+  this->item = std::move(item);
+  this->responses = std::move(responses);
+  this->deterministic = true;
+  ResetDeterministic();
+
+  if (!reset)
+    ResetParameters();
+}
+
+template<typename Model, typename InitializationRuleType>
+template<typename OptimizerType>
+void NCFNetwork<Model, InitializationRuleType>::Train(arma::mat user,
+                                                      arma::mat item,
+                                                      arma::mat responses,
+                                                      OptimizerType& optimizer)
+{
+  ResetData(std::move(user), std::move(item), std::move(responses));
+
+  // Train the model.
+  Timer::Start("ncf_optimization");
+  const double out = optimizer.Optimize(*this, parameter);
+  Timer::Stop("ncf_optimization");
+
+  Log::Info << "NCFNetwork::NCFNetwork(): final objective of trained model is "
+      << out << "." << std::endl;
+}
+
+template<typename Model, typename InitializationRuleType>
+void NCFNetwork<Model, InitializationRuleType>::Forward(arma::mat userInput,
+                                                        arma::mat itemInput,
+                                                        arma::mat& results)
+{
+  if (parameter.is_empty())
+    ResetParameters();
+
+  if (!deterministic)
+  {
+    deterministic = true;
+    ResetDeterministic();
+  }
+
+  curUserInput = std::move(userInput);
+  curItemInput = std::move(itemInput);
+  Forward(std::move(curUserInput), std::move(curUserInput));
```

Comment:
You are right. The merged (multiplymerge/concat) output will further pass through two more layers, to give final output. If the existing classes can create this network, that will be great. Pseudo code will be much helpful. Thanks.

Reply:
So one idea I had was to implement a subview layer, which takes the merged input (user/item) and only selects the interesting part for a particular network, something like:

```c++
FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization> modelA;
modelA.Add<Subview<> >(0, 10); // here we take the first 10 elements from the input
modelA.Add<IdentityLayer<> >();
modelA.Add<LinearNoBias<> >(10, 2);

FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization> modelB;
modelA.Add<Subview<> >(10, 20); // here we take the last 10 elements
modelB.Add<IdentityLayer<> >();
modelB.Add<LinearNoBias<> >(10, 2);

FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization> modelMain;
modelB.Add<MergeMultiply<> >(modelA, modelB);
```

in the ``Train`` function we can merge both inputs and use the existing functionalty as before.

Another idea is to avoid an extra subview layer and to do the merge of the output inside the ``NCF`` class. I guess this is what you currently doing, let me know what you think about each idea.

Reply:
The first idea does solve the problem of passing multiple input matrices. But there was an issue with using `multiplymerge` in `ffn`. We had considered using `sequential` instead but this didn't work out when trying to add further layers after `multiplymerge`(linear, sigmoid) and was giving 'double free or corruption' error. Unless I am implementing the network wrong, this will also  have to be sorted out to implement the first idea.
With the second one, you are right. It is now based on merging outputs in `NCF` class, but even here I'm concerned with adding further layers after the merging part (since it is manual merging, and not part of the network) to form a continuous network.
I think first one is a better approach since the training problem will be completely solved and no manual merging is required. Just need to resolve that error.

Reply:
Okay, I'll take a look into the sequential issue tomorrow.

Reply:
I found a couple of issues, which needs to be fixed, will open a PR, in the meantime do you like to work on the subview layer?

Reply:
Sure, that sounds like a good plan. Shall I close this PR then? I can just open another one for subview layer.

Reply:
Let's keep this one open, once the changes are merged we can reuse this one. For the subview layer let's open separate PR.

Reply:
Here are the modifications: https://github.com/mlpack/mlpack/pull/1427, this one comes with a simple example, let me know what you think (The subview layer would be the first layer of the Sequential layer).